### ISSUE TITLE
Comp and Pen Scheduling Notification Sends

### ISSUE THREAD

**Issue by EvanParish** [department-of-veterans-affairs/notification-api#1506]
## User Story - Business Need
The goal is to send out all the Comp and Pen notifications at the end of the month. The best approach seems to be a celery task running on a cron in the API. This task can grab the data from the Dynamo DB cache table and send the notifications in batches of `x/minute` using cron to trigger every minute during our send window. Doing this directly in the API will allow us to bypass our auth checks and reduce the overhead necessary to process these requests.

- [x] Ticket is understood, and QA has been contacted (if the ticket has a QA label).

### User Story(ies)

**As a** VA Notify
**I want** create a task to send notifications 
**So that** we can send Comp and Pen notifications out in a reasonable amount of time at the end of each month

### Additional Info and Resources

[Architecture Diagram](https://github.com/department-of-veterans-affairs/vanotify-team/blob/master/Architecture/comp%20&%20pen/readme.md#bip-integration-architecture) 
[Celery cron scheduling](https://docs.celeryq.dev/en/stable/reference/celery.schedules.html)  

Example of how to send notifications within our code:

[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBMloxQVE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--40246b4a10f506a39e68be8b3d237bea7165436f/image.png)

Time to send in the following table is based on sending 7 mil messages. Please note - we cannot exceed 100 m/s due to short code throughput limits.

|msgs / min|msgs / sec|hrs to send all msgs|# 8 hr days|
|---|---|---|---|
|1,000|16.67|116.67|14.58|
|2,000|33.33|58.33|7.29|
|3,000|50.00|38.89|4.86|
|4,000|66.67|29.17|3.65|
|5,000|83.33|23.33|2.92|
|6,000|100.00|19.44|2.43|
|7,000|116.67|16.67|2.08|
|8,000|133.33|14.58|1.82|
|9,000|150.00|12.96|1.62|
|10,000|166.67|11.67|1.46|

## Engineering Checklist

- [x] Create a celery task that will pull 3000 messages from the dynamo db table, prepare the data, and queue a notification to be sent (bypassing the need to call an endpoint)
  - [ ] The table query/scan grabs 3000 that have "processed == False"
  - [x] Method calls `send_to_queue_for_recipient_info_based_on_recipient_identifier()`
  - [x] After sending the notification to the queue, the table record is updated with a "processed == True" property
- [x] Create Celery Beat cron in the [beat_schedule](https://github.com/department-of-veterans-affairs/notification-api/blob/master/app/config.py#L264) to run the celery task every minute between 1 PM and 9 PM Eastern time
  -  The cron needs to run after the cache table is populated from the mpi lambda and the kafka consumer lambda
  -  scheduling a celery task can be done in `app/Config.py` under `CELERY_SETTINGS` 
  -  should look something like this: `* 13-21 21-28 * *` - **note:** you could add `1-5` for last section to only send on weekdays 
      -   **Description:** _At every minute past every hour from 13 through 21 on every day-of-month from 21 through 28._ 
      -   **Note:** our app is set to _Eastern_ time, so the expression should assume it's running in _Eastern_ time. 
      -   **Note:** The cron needs to activate after we have finished updating the cache table for the month and all the messages from the kafka consumer have been placed in the table
- [x] Set up proper [logging](https://github.com/department-of-veterans-affairs/vanotify-team/blob/master/Engineering/logging_policy.md)
- [ ] Celery Beat can handle processing 3000 requests in a minute (try making 3000 and ensure it finishes in time - time documented as a comment)
  - This will be done on a follow-up ticket
  - Take a screenshot of how long the task took to run such as:

[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBK0oxQVE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--49ae5f511fe19c552f3da78cc1ac3205efed6093/image.png)

## Functional Testing
- [x] Manually add to the dynamodb table before the cron starts
  - [x] Nothing is processed until the expected time
- [x] Tweak the Celery Task to only pull 1/minute, then add several to the dynamodb table before the end of the cron, ensuring it stops when it is supposed to
  - [x] Nothing is processed after the end time
- [x] Table records processed are sent (validation through observation)

## Acceptance Criteria

- Notifications are only sent within the defined cron schedule
- SMS notifications are successfully delivered

## QA Considerations
_For QA to populate. Leave blank if QA is not applicable on this ticket._

- QA is included for awareness. Logs are going to have to be checked to verify this runs correctly.

## Potential Dependencies
_Leave blank if n/a_
~To fully test this is integrated with the rest of the process the DynamoDB will need to be set up with kafka events~
This only requires that the table is populated as-necessary

## Out of Scope
Do not rewrite functionality. Prepare the notification and utilize `send_to_queue_for_recipient_info_based_on_recipient_identifier()`

**Comment by mjones-oddball** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/VA-Notify-620d21369d810a00146ed9c8/issues/department-of-veterans-affairs/notification-api/1506?planning-poker) with Zenhub @babyjunior56 @cris-oddball @EvanParish @k-macmillan @kalbfled @ldraney @nikolai-efimov

**Comment by cris-oddball** [comment]
@EvanParish a reminder that this ticket needs updates based on Weds refinement session before start of business on Monday so that we can refine it then. Can you please ping me when that is complete?

**Comment by cris-oddball** [comment]
Kyle has updated this ticket.

**Comment by EvanParish** [comment]
I made good progress with testing the code today. I verified that we can connect to the dynamodb table from the app without any infra changes needed. I successfully pulled info from the table and if I had valid vaprofile_id values I believe that notifications would send. I ran into an issue with how the limit works when grabbing data from the dynamodb, but that was overcome. The last piece I need to verify before I call this done is that I understand how updating the items in the table works. I'm currently getting an error when trying to do so.

**Comment by EvanParish** [comment]
I was hoping I'd have the PR ready for review today, but after talking with Cris we identified some more problems that need to be solved, like the phone number not being attached to the notification when trying to send it. I'm looking into it and will hopefully have the PR up tomorrow before I head out.

**Comment by cris-oddball** [comment]
Just noting this won't make it into the next deployment.

**Comment by EvanParish** [comment]
I got the `sms_sender_id` populating, but I'm still missing the `phone_number`/`to` field. The PR will definitely not be up before heading out today. I'm still working on figuring this phone number issue out.

Here's the latest notification send attempt.
```
{
    "billing_code": null,
    "body": "Comp and Pen test message with amount: 111",
    "completed_at": "2023-11-09T18:30:13.425550Z",
    "cost_in_millicents": 0.0,
    "created_at": "2023-11-09T18:30:13.136736Z",
    "created_by_name": null,
    "email_address": null,
    "id": "caa08793-273f-40db-9985-2d22077cf3f8",
    "phone_number": null,
    "postage": null,
    "postcode": null,
    "provider_reference": null,
    "recipient_identifiers": [
        {
            "id_type": "VAPROFILEID",
            "id_value": "112491"
        }
    ],
    "reference": null,
    "scheduled_for": null,
    "segments_count": 0,
    "sent_at": null,
    "sent_by": null,
    "sms_sender_id": "286d6176-adbe-7ea7-ba26-b7606ee5e2a4",
    "status": "technical-failure",
    "status_reason": "VA Notify non-retryable technical error",
    "subject": null,
    "template": {
        "id": "fca415e8-ff5f-4c29-a121-6dd534aa90c7",
        "uri": "https://dev-api.va.gov/v2/template/fca415e8-ff5f-4c29-a121-6dd534aa90c7/version/1",
        "version": 1
    },
    "type": "sms"
}
```

**Comment by EvanParish** [comment]
After a quick chat with Kyle and an easy fix the notifications are now sending! Monday morning I'll clean up the code and get a PR out for review.

**Comment by EvanParish** [comment]
Unit tests are being very difficult, so this is carrying over into the next sprint at this point. They should take too much longer, but the PR is large so reviewing will take some time.

**Comment by EvanParish** [comment]
Finally managed to get all the unit tests in order and now I'm just waiting on reviews for the PR and addressing changes as they come up.

**Comment by EvanParish** [comment]
I have made all the requested changes. Awaiting final approval. If it's approved while I'm out @cris-oddball will merge it.

**Comment by cris-oddball** [comment]
The PR has two approvals. I will be merging it this morning.

**Comment by cris-oddball** [comment]
Merged but deploy to Perf failed because the perf celery tasks had the new parameters in the wrong place ('secrets' instead of 'environments'). Kyle to put up a fix for it since Evan is out.

**Comment by cris-oddball** [comment]
Hotfix was deployed to Perf, the code deployed successfully and is now merged to master. Closing ticket.
